ale_linters/sh/language_server.vim

```diff
@@ -0,0 +1,34 @@
+" Author: Christian Höltje (https://docwhat.org/)
+" Description: BASH Language server integration for ALE
+scriptencoding utf-8
+
+call ale#Set('sh_language_server_executable', 'bash-language-server')
+call ale#Set('sh_language_server_use_global', get(g:, 'ale_use_global_executables', 0))
+
+function! ale_linters#sh#language_server#GetExecutable(buffer) abort
+    return ale#node#FindExecutable(a:buffer, 'sh_language_server', [
+    \   'node_modules/.bin/bash-language-server',
+    \   'bash-language-server',
+    \])
+endfunction
+
+function! ale_linters#sh#language_server#GetCommand(buffer) abort
+    let l:exe = ale#Escape(ale_linters#sh#language_server#GetExecutable(a:buffer))
+
+    return l:exe . ' start'
+endfunction
+
+function! ale_linters#sh#language_server#GetProjectRoot(buffer) abort
+    let l:git_path = ale#path#FindNearestDirectory(a:buffer, '.git')
+
+    return !empty(l:git_path) ? fnamemodify(l:git_path, ':h:h') : '.'
```

Comment:
Be aware that `'.'` will start one server per file on the system, and it will keep running until Vim closes or you use `:ALEStopAllLSPs`.

Reply:
Is there something better I should do to handle files that aren't in a "project"?

I'm still new to this whole LSP thing.

Reply:
I suppose it could use some arbitrary directory for the project, like `/`.

Reply:
I honestly don’t understand what the root is for.

Reply:
> Be aware that '.' will start one server per file on the system, and it will keep running until Vim closes or you use :ALEStopAllLSPs.

Is that one server per _root_ on the system?  e.g. If you edit different files in the same directory would it be one server?   Or is it really per-file?

What happens if the file is outside the _root_?  e.g. if we use `$HOME` and the file is in `/etc/`?

I could write some code that looks if the file is in `$HOME` and use that or else use `/`.

Reply:
The issue is that ALE runs one instance of an LSP server per project, and you typically have to establish where a project is. I suppose `'.'` might so happen to work where the path is kind of ignored and one server is started for all files not in git repositories. LSP is designed to work with projects which have some project root path, instead of individual files spread across the file system.

For files which aren't in any project in particular, the best thing I can think of is just to set the project root to be the root of the filesystem the file is in.